README.md
# Noko Client

[![Documentation Status](https://readthedocs.org/projects/python-freckle-client/badge/?version=latest)](https://python-freckle-client.readthedocs.io/en/latest/?badge=latest)

A simple Noko (formerly Freckle) API client implementation. Offers two different clients with different levels of flexibility
to interact with the Noko API v2.

Full documentation on the NokoClient can be found [here](https://python-freckle-client.readthedocs.io/en/latest/).

<!-- HIDE_INTRO -->
## Installation

To get the latest stable release from PyPi:

```shell
pip install python-freckle-client
```

**Requirements:**

Version `v0.5.0` and lower requires at least Python 3.8. Version `v1.0.0` and upper require Python 3.10 and Pydantic v2.

## Usage

The package offers three different clients:

- NokoClient (Noko's v2 API)
- FreckleClientV2 (Noko's v2 API)
- FreckleClient (Noko's v1 API) <- Deprecated!

To use the `NokoClient`, just import the client, create an instance and call the desired method:

```python
from noko_client.client import NokoClient

client = NokoClient('access_token')
entries = client.list_entries(from_="2023-08-01", to=datetime(2023, 8, 15))
```

To use the `FreckleClient` or the `FreckleClientV2`, just import the client, create an instance and call the `fetch_json` method: 

```python
from freckle_client.client import FreckleClient

client = FreckleClient('account_name', 'api_token')
entries = client.fetch_json(
    'entries',
    query_params={
        'per_page': 1000,
        'search[from]': '2015-01-01',
        'search[to]': '2015-01-31',
        'search[projects]': [1423, 24545, ],
    }
)
```

Similarly, to use the V2 API:

```python
from freckle_client.client import FreckleClientV2

client = FreckleClientV2('access_token')
entries = client.fetch_json(
    'entries',
    query_params={
        'from': '2015-01-01',
        'to': '2015-01-31',
        'billable': 'true',
        'project_ids': '12345,67890'
    }
)
```


## Contribute

If you want to contribute to this project, please fork this repository and clone your fork, then set up a virtual environment:

```shell
virtualenv -p 3.10 venv
source venv/bin/activate
pip install -r requirements-dev.txt
```

Create a new branch, implement your feature or fix, and send us a pull request.


CHANGELOG.txt
=== 1.1.0 ===

Add package as py.typed

=== 1.0.5 ===

Minor fixes to docs and to the Noko client

=== 1.0.0 ===

* Implement new NokoClient for the v2 API
* Support individual methods for all resource endpoints
* Support parameter validation on requests
* Support passing parameters as keyword arguments and as natural types

=== 0.5.0 ===

* Handle successful No Content response
* Handle pagination on the v2 API

=== 0.4.0 ===

* Uses new application name & URL (nokotime.com)

=== 0.3.X ===

* prepared app for Python3.x
* Fixed request headers
* Fixed response handling

=== 0.2 ===

* Raising proper HTTP exception from the Requests library instead of our
  self-made FreckleClientException if the request was not successful.
* IMPORTANT: If you were catching FreckleClientException before, please make
  sure to catch `requests.exceptions.HttpError` now.

=== 0.1 ===
- Initial commit


# Suggested file syntax:
#
# === (ongoing) ===
# - this is always on top of the file
# - when you release a new version, you rename the last `(ongoing)` to the new
#   version and add a new `=== (ongoing) ===` to the top of the file
#
# === 1.0 ===
# - a major version is created when the software reached a milestone and is
#   feature complete
#
# === 0.2 ===
# - a minor version is created when a lot of new features have bene added or
#   significant backwards incompatible changes have been made.


freckle_client/client.py
"""
Freckle client implementation.

Idea stolen from here:
https://github.com/sarumont/py-trello/blob/master/trello/__init__.py#L108

"""
import json
import warnings
from typing import Callable

import requests

from . import __version__, exceptions


def deprecated(func: Callable) -> Callable:
    """Decorate a function to flag a method as deprecated."""

    def deprecated_func(*args, **kwargs) -> Callable:  # noqa: ANN002, ANN003
        warnings.warn(
            "The FreckleClient is deprecated. Please use the NokoClient or the FreckleClientV2 instead.",
            category=DeprecationWarning,
            stacklevel=2,
        )
        return func(*args, **kwargs)

    return deprecated_func


class FreckleClient:
    """Simple client implementation to fetch json data from the v1 API."""

    def __init__(self, account_name: str, api_token: str):
        """
        Create a ``FreckleClient`` instance.

        :account_name: Your Freckle account name.
        :api_token: Your Freckle API token.

        """
        self.account_name = account_name
        self.api_token = api_token

    @deprecated
    def fetch_json(
        self,
        uri_path: str,
        http_method: str = "GET",
        headers: dict | None = None,
        query_params: dict | None = None,
        post_args: dict | None = None,
    ) -> list[dict] | None:
        """
        Fetch some JSON from Noko.

        For example, fetch some entries like so:

            entries = self.fetch_json(
                'entries',
                query_params={
                    'per_page': 1000,
                    'search[from]': '2015-01-01',
                    'search[to]': '2015-01-31',
                    'search[projects]': [1423, 24545, ]),
                }
            )

        """
        # explicit values here to avoid mutable default values
        if headers is None:
            headers = {}
        if query_params is None:
            query_params = {}
        if post_args is None:
            post_args = {}

        # set content type and accept headers to handle JSON
        headers["Accept"] = "application/json"
        query_params["token"] = self.api_token

        # construct the full URL without query parameters
        url = f"https://{self.account_name}.nokotime.com/api/{uri_path}.json"

        # perform the HTTP requests, if possible uses OAuth authentication
        response = requests.request(
            http_method,
            url,
            params=query_params,
            headers=headers,
            data=json.dumps(post_args),
        )

        if response.status_code != 200:
            raise exceptions.FreckleClientException(
                "Freckle API Response is not 200", response.text
            )

        # return content if successful response with content,
        # otherwise return None
        return json.loads(response.content) if response.content else None


class FreckleClientV2:
    """Simple client implementation to fetch json data from the v2 API."""

    def __init__(self, access_token: str):
        """
        Create a ``FreckleClient`` instance.

        :account_name: Your Freckle account name.
        :api_token: Your Freckle API token.

        """
        self.access_token = access_token

    def fetch_json(
        self,
        uri_path: str,
        http_method: str = "GET",
        headers: dict | None = None,
        query_params: dict | None = None,
        post_args: dict | None = None,
    ) -> list[dict] | None:
        """
        Fetch some JSON from Noko.

        For example, fetch some entries like so:

            entries = self.fetch_json(
                'entries',
                query_params={
                    'search[from]': '2015-01-01',
                    'search[to]': '2015-01-31',
                    'search[projects]': [1423, 24545, ]),
                }
            )

        """
        # explicit values here to avoid mutable default values
        if headers is None:
            headers = {}
        if query_params is None:
            query_params = {}
        if post_args is None:
            post_args = {}

        # set content type and accept headers to handle JSON
        headers["Accept"] = "application/json"
        headers["User-Agent"] = f"python-freckle-client/{__version__}"
        headers["X-FreckleToken"] = self.access_token

        # construct the full URL without query parameters
        url = f"https://api.nokotime.com/v2/{uri_path}"
        response = self._make_request(
            http_method, url, headers, query_params, post_args
        )
        # return content if successful response with content,
        # otherwise return None
        return response

    # private

    @staticmethod
    def _make_request(
        http_method: str,
        url: str,
        headers: dict | None = None,
        query_params: dict | None = None,
        post_args: dict | None = None,
    ) -> list[dict] | None:
        results: list = []
        while url:
            # perform the HTTP requests, if possible uses OAuth authentication
            response = requests.request(
                http_method,
                url,
                params=query_params,
                headers=headers,
                data=json.dumps(post_args),
            )
            # if request failed (i.e. HTTP status code not 20x),
            # raise appropriate error
            response.raise_for_status()

            if not response.content:
                return None

            results.extend(response.json())
            next_link = response.links.get("next")

            if not next_link:
                break

            url = next_link["url"]

        return results
